### ISSUE TITLE
[Feature Pitch] Fast extremal eigensolvers

### ISSUE THREAD

**Issue by rfeinman** [pytorch/pytorch#58828]
## 🚀 Feature

When working with large matrices, it's useful to access one or a handful of desired eigenpairs without computing the complete spectrum. PyTorch currently offers only `torch.lobpcg`, which is limited to positive-definite matrices and is very slow in my experience. I've written a [PyTorch C++ extension](https://github.com/rfeinman/Torch-ARPACK) for identifying extremal eigenpairs based on canoncial Arnoldi & Lanczos techniques. These tools are applicable to a wide range of ML problems and would be a useful addition to the core PyTorch library. The algorithms are based on matrix-vector products and are therefore relevant to the ongoing `torch.sparse` project.

## Motivation

Regarding current PyTorch offerings, `torch.lobpcg` is limited to positive-definite matrices and it does not appear to be an efficient implementation of the LOBPCG algorithm (much slower than SciPy's equivalent lobpcg on CPU). `nn.utils.spectral_norm` computes a single eigenpair using power iteration; however, power iteration has poor convergence properties, and in the special case of symmetric/hermitian matrices, convergence can be easily improved with little trade-off.

The motivation is to augment the PyTorch library with fast, compiled tools based on cannonical algorithms for extremal eigensolvers. The proposed package is inspired by the [ARPACK](https://www.caam.rice.edu/software/ARPACK/) Fortran library. It uses only matrix-vector products and is therefore applicable to sparse matrices. Even for matrices stored in dense format, the algorithm is faster and has a smaller memory footprint compared to full-spectrum methods.

## Pitch

The current package (temporarily named "arpack") includes a single tool `eigsh` which is applicable to symmetric matrices that are potentially indefinite. At each step, the algorithm performs an m-step Lanczos factorization and subsequently calls LAPACK routines *?stebz* and *?stein* to compute a partial Schur decomposition of the resulting tridiagonal system. There are no MAGMA analogues to *?stebz* and *?stein* that I'm aware of; therefore, only CPU is currently supported (see "Alternatives" below for some proposals re:GPU).

Snapshot:
```python
import torch
import arpack

# symmetric matrix input
A = torch.randn(1024, 1024)
A = torch.mm(A.t(), A)

# quickly compute the largest/smallest (algebraic) eigenpair
eigval, eigvec, n_iter = arpack.eigsh(A, largest=True)
```

In my runtime expriments, the algorithm is much faster than `torch.linalg.eigh`/`torch.lobpcg` and it comes close to `scipy.sparse.linalg.eigsh`. These experiments were run on CPU using a pytorch compiled with MKL and OpenMP threading.

Results:

```
[--------------------------- eigensolvers --------------------------]
                        |   256    |   512    |    1024   |    2048
6 threads: ----------------------------------------------------------
      torch_eigh        |  1803.1  |  7187.1  |  30054.7  |  208623.6
      torch_lobpcg      |  5670.4  |  9149.9  |  38405.4  |   47878.6
      arpack_eigsh      |   366.0  |  2171.9  |   4787.2  |   15731.7
      scipy_eigsh       |   629.9  |  1023.1  |   3104.4  |   11026.9

Times are in microseconds (us).
```

The benchmark script can be viewed [here](https://github.com/rfeinman/Torch-ARPACK/blob/master/scripts/benchmark.py).

## Alternatives

In addition to the custom `eigsh` function, I've also included a function `eigsh_mkl` that wraps an equivalent tool from the Intel Math Kernel Library (MKL). It is a bit slower than my own implementation, but potentially more reliable as an Intel-backed product.

I've started tracking a list of relevant CPU and CUDA implementations for extremal eigenvalue problems [here](https://github.com/rfeinman/Torch-ARPACK#appendix-cpu--cuda-partial-eigensolver-index). For positive definite matrices, there appears to be an implementation of LOBPCG available for both platforms. In general, offerings on CUDA are relatively limited. My own `eigsh` implementation is portable to CUDA in theory, with the exception of LAPACK routines *?stebz* and *?stein*, for which I'm not aware of any MAGMA/CuSOLVER alternatives.

**Comment by vadimkantorov** [comment]
cc @lobpcg

**Comment by lobpcg** [comment]
@vadimkantorov 
lobpcg is NOT limited to positive-definite matrices. It works for generalized eigenvalue problems Ax=lambda Bx, where B must be  positive-definite, which is trivially the case if B is the identity, but the matrix A is NOT limited to positive-definite. 
I would be very surprised if specifically torch.lobpcg is limited to positive-definite matrices A.

**Comment by lobpcg** [comment]
@vadimkantorov 
If torch.lobpcg does not appear to be an efficient implementation of the LOBPCG algorithm, e.g., much slower than SciPy's equivalent lobpcg on CPU, may be a bug report is in order. At any rate, it would be helpful if you add SciPy's lobpcg numbers to your table, just as for scipy_eigsh

**Comment by pearu** [comment]
> torch.lobpcg is limited to positive-definite matrices and it does not appear to be an efficient implementation of the LOBPCG algorithm (much slower than SciPy's equivalent lobpcg on CPU).

Note that
```
pytest -sv test/test_linalg.py -k test_lobpcg_scipy
```
compares the timings and robustness of scipy and torch lobpcg implementations:
```
CPU timings: torch.lobpcg vs scipy.sparse.linalg.lobpcg
-------------------------------------------------------
              | standard    | generalized | method
torch.lobpcg  |     124.37  |     147.61  | ortho
scipy_lobpcg  |     114.87  |     122.50  | N/A
-(input size:  500, eigenpairs: 7, units: ms per call)-

Handling of small tol=1e-100: torch.lobpcg vs scipy.sparse.linalg.lobpcg
----------------------------------------------------------------------------
              | standard    | generalized |  niter | method
torch.lobpcg  |   1.14e-13  |   8.24e-15  |   1000 | ortho
scipy_lobpcg  |   1.53e-07  |   3.71e-10  |   1002 | N/A
---(input size:  500, eigenpairs: 7, units: relative error, maxiter=1000)---
```
These results show that
- torch.lobpcg is not worse than scipy lobpcg with respect to robustness
- torch.lobpcg is slightly slower but this has an explanation: torch.Tensor is a "heavier" object than numpy ndarray and for small sizes object creation will have a noticeable overhead.

Could you provide more test data to back up your motivation statement?

**Comment by Lezcano** [comment]
@rfeinman when you provide the tests, could you please do it against torch master / nightly? We added a number of improvements to `torch.linalg` in this last release, so there are some chances that `eigh` benefited of these.

**Comment by rfeinman** [comment]
@lobpcg Yes, I applogize, you are right about the LOBPCG algorithm in general. However, in my experience, `torch.lobpcg` produces incorrect results for non-PD matrices. Consider the following example.

Code (using sample_symmetric from [benchmark.py](https://github.com/rfeinman/Torch-ARPACK/blob/master/scripts/benchmark.py)):

```python
import torch
from scipy.sparse import linalg as splinalg

torch.manual_seed(491)
A = sample_symmetric(512, dtype=torch.float64)

E, V = torch.linalg.eigh(A)
print('eig0 (true): {:.4f}'.format(E[0]))

e, _ = torch.lobpcg(A, k=1, largest=False, tol=1e-4)
print('eig0 (torch_lobpcg): {:.4f}'.format(e.item()))

X = A.new_empty(A.size(0), 1).normal_()
e, _ = splinalg.lobpcg(A.numpy(), X.numpy(), largest=False, tol=1e-4)
print('eig0 (scipy_lobpcg): {:.4f}'.format(e.item()))
```

Produces:

```
eig0 (true): -28.0466
eig0 (torch_lobpcg): -13.9858
eig0 (scipy_lobpcg): -28.0466
```

The same result occurs with method="basic", and different values of tol.

@Lezcano I am aware of the improvements to torch.linalg (nice work btw!). I'm using a very recent Git pull for the tests reported above. I built from source because the macOS conda binaries don't use multi-threaded BLAS (headache). Eigh is slower with the nightly mac binary.

I will share test results comparing torch and scipy lopcg runtime shortly. However, my comment about the differential with scipy was only a side point. More importantly, you'll notice that torch.lopcg is slower than full eigendecomposition (torch.linalg.eigh) even with a 1024x1024 matrix. In and of itself, that seems to call for a new tool.

**Comment by rfeinman** [comment]
@pearu @lobpcg I added results with scipy lobpcg to the benchmark table in my original ticket post. The updated script is available [here](https://github.com/rfeinman/Torch-ARPACK/blob/master/scripts/benchmark.py).

Notice the large improvement vs. `torch.lobpcg`, especially for n=1024, 2048.

**Comment by lobpcg** [comment]
scipy_eigsh is the scipy interface to arpack, and it gives similar timing to the proposed code, as expected. 

The comparison is thus basically Lanczos vs. LOBPCG. For small size dense random matrices selected in these tests for a single eigenvalue the results are plausible, primarily due to a faster convergence of Lanczos and the need to multiply by the dense matrix.

**Comment by rfeinman** [comment]
@lobpcg as far as I understand, any algorithm based on matrix-vector products should get a speedup from sparse vs. dense matrices. Are you suggesting that the speedup will be greater for LOBPCG compared to these alternatives?

**Comment by pearu** [comment]
Notice that
```python
>>> e, _ = torch.lobpcg(-A, k=1, largest=True, tol=1e-4)
>>> -e
tensor([-28.0466], dtype=torch.float64)
```
So, something goes wrong when `largest=False` is used: the eigenvalues are not sorted when using `k>1`. It could be a bug.

Btw, it looks like the benchmark script uses torch.linalg.eigh to compute all eigenvalues while for other methods, only the largest eigenvalue is acquired. This may explain the high timing value for `torch_eigh`.

One should also realize that different algorithms have different domains of applications, in general. It makes more sense to directly compare the timings of different implementations of the same algorithm. Comparing two algorithms has to be done more carefully, for instance, tuning of algorithm parameters may change the timings and accuracy drastically.

Finally, I think the used input sizes (that are in the order of 10^3) are rather small and it may be more efficient to use some full solver on a dense input rather than using the LOBPCG method, for instance, which is designed for much larger (sparse) inputs, say, with sizes in the order of 10^6 or more.

I think the overall goal of adding more solvers to torch is fine.

Re motivation: from my experience, I have not seen yet the evidence for justifying the usage superlatives on how bad is torch.lobpcg or the LOBPCG method in general. If such evidence exists, an analysis that pinpoints the bottlenecks would be of interest.

**Comment by lobpcg** [comment]
For a single eigenvalue, the entire cost in lobpcg basically goes into matvec. Not so for Lanczos. Try running on diagonal matrices and always report the number of iterations.



________________________________

**Comment by rfeinman** [comment]
@pearu yes, I'm aware that torch.linalg.eigh computes the complete spectrum. I included it primarily because, other than torch.lobpcg, there is no other way to compute a subset of eigenpairs in pytorch.

I understand that for smaller matrices we can simply call torch.linalg.eigh without worrying too much about runtime. It creates the burden of having to write programs that check the matrix size and decide which function to call accordingly. I was motivated by the following desirable qualities of SciPy's `eigsh`:
1. The algorithm will capitalize on sparsity when it is available, and will also do just fine when it is not.
2. The algorithm is always _at least as fast_ as eigh, never slower.

Together, these seem like the most sensible and desirable qualities for an extremal eigensolver.

**Comment by rfeinman** [comment]
@pearu I've put together some more extensive tests to demonstrate the shortcomings of PyTorch lobpcg vs SciPy in the large, sparse setting. Apologies for the delay - it took me a while to put the code together.

As a preface: I've noticed that the `torch.sparse` module is fairly limited. The stable release includes only COO format, which is not ideal for matvec and is rarely used in sparse eigensolvers. CSR was introduced recently, although only for Windows & Linux, and CUDA is not supported. And unlike SciPy, there is no way to define custom sparse operators in PyTorch (as e.g. LinearOperator).

Therefore, although @pearu and @lobpcg point out that torch.lobpcg is meant to be used with large, sparse matrices, there are few ways for pytorch users to take advantage of lobpcg in this setting. Constructing sparse matrices is impractical in pytorch: there are no constructor utilities like scipy's sparse.diags and sparse.block_diag, so users need to hand-specify indices (painful), or use dense constructors followed by `.to_sparse()` (infeasible for order 10^6 matrices on most machines). These items considered, I think my findings already with dense matrices are relevant and important.

__Sparse LOBPCG Benchmark__

I've put together the following benchmark script, which can be run using a recent nightly pytorch binary for Windows or Linux: [benchmark_lobpcg.py](https://github.com/rfeinman/Torch-ARPACK/blob/master/scripts/benchmark_lobpcg.py).

The script generates random matrices with a handful of different sparsity structures (e.g. diagonal, block-diagonal, block-toeplitz). Each structure is tested in sparse-CSR format at various sizes. I've taken care to ensure that equivalent sparse formats are used for both scipy_lobpcg and torch_lobpcg. For PyTorch, I'm using int32 CSR indices to get fast, MKL-based matvec ops as per the [documentation tip](https://pytorch.org/docs/master/sparse.html#sparse-csr-tensor).

Using a recent nightly build on my Linux machine with 16 threads, I get the following results:

```
[------------------ diag (k=5, largest=True) -----------------]
                    |  1.0240e+03  |  1.6384e+04  |  2.6214e+05
16 threads: ---------------------------------------------------
      torch_lobpcg  |    2438.5    |   11534.6    |   19820.8
      scipy_lobpcg  |      19.6    |     202.3    |    1917.4

Times are in milliseconds (ms).


[--------------- block_diag (k=5, largest=True) --------------]
                    |  1.0240e+03  |  1.6384e+04  |  2.6214e+05
16 threads: ---------------------------------------------------
      torch_lobpcg  |    1595.2    |    8533.3    |   44303.7
      scipy_lobpcg  |      55.2    |     657.0    |    8316.4

Times are in milliseconds (ms).
```

As you can see, the results with large sparse matrices are very similar to my original tests: scipy's lobpcg is 5-15x faster than torch on average, and this replicates across different orders.

I've only run the experiment for k=5 and largest=True so far. However, I've written the script for anyone to use if you'd like to extend the tests. The only requirements are python, scipy and pytorch.

**Comment by lobpcg** [comment]
@rfeinman do you reproduce the results of 
https://github.com/pytorch/pytorch/issues/58828#issuecomment-847545925 in your  setup?

**Comment by rfeinman** [comment]
@lobpcg great suggestion. When I tried just now I was getting different results at first. I updated to the 1.9.0.dev20210601 binary and now I'm reproducing the test results closely. I also re-ran my own lobpcg benchmark and got different results - see section "update" of my https://github.com/pytorch/pytorch/issues/58828#issuecomment-852610307.

results from `pytest -sv test/test_linalg.py -k test_lobpcg_scipy`:

```
test_linalg.py::TestLinalgCPU::test_lobpcg_scipy_cpu_float64
CPU timings: torch.lobpcg vs scipy.sparse.linalg.lobpcg
-------------------------------------------------------
              | standard    | generalized | method
torch.lobpcg  |     171.58  |     211.95  | ortho
scipy_lobpcg  |     138.29  |     134.94  | N/A
-(input size:  500, eigenpairs: 7, units: ms per call)-

Calling scipy_lobpcg failed [standard]: eigh has failed in lobpcg iterations
Handling of small tol=1e-100: torch.lobpcg vs scipy.sparse.linalg.lobpcg
----------------------------------------------------------------------------
              | standard    | generalized |  niter | method
torch.lobpcg  |   1.92e-13  |   8.47e-15  |   1000 | ortho
scipy_lobpcg  |  -1.00e+00  |   1.41e+01  |     -1 | N/A
---(input size:  500, eigenpairs: 7, units: relative error, maxiter=1000)---
```

One difference is that scipy seems to fail in the 2nd test for me.

**Comment by mollenhauerm** [comment]
Bump: has there been any progress on this since the issue was opened in 2021?
